### PR TITLE
ISPN-5420 Thread pools are depleted by ClusterTopologyManagerImpl.waitFo...

### PR DESCRIPTION
**Pull request by danberindei**
...rView() and causing deadlock

https://issues.jboss.org/browse/ISPN-5420

* LargeCluster2StressTest: New stress test with fewer nodes but more caches than LargeClusterStressTest
* Better logging of command responses
* Add a queue for the state transfer executor (embedded only for now)
* Fix SemaphoreCompletionService.continueTaskInBackground(null)
* Move all StateRequestCommands to the remote executor thread pool
* Check for tasks to execute immediately after installing the topology,
  because StateRequestCommands don't wait for the transaction data
* Don't send a rebalance confirmation if we were not a member
* In case the state transfer executor has a caller runs policy, don't keep
  the transfer maps lock while submitting the segments request task
* Undo the SemaphoreCompletionService for view handling in ClusterTopologyManagerImpl
* Update the cache members without holding the view handling lock
* Use a SemaphoreCompletionService per cache to handle topology updates in
  LocalTopologyManagerImpl

### CODE REVIEW

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java [hunk 1]
```diff
@@ -301,6 +301,7 @@ public void onTopologyUpdate(final CacheTopology cacheTopology, final boolean is
       }
       stateTransferLock.releaseExclusiveTopologyLock();
       stateTransferLock.notifyTopologyInstalled(cacheTopology.getTopologyId());
+      remoteCommandsExecutor.checkForReadyTasks();
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: haven't we decided that this is bad (and the line 390 too) and we should move it after the command is processed?

Comment (danberindei) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: Unfortunately after the command is processed also means after the cluster listeners and the transaction data were received, which might be too late. But you're right, in my test I didn't have any write operations that could fill the remote executor and cause `checkForReadyTasks()` to block.

I'll try again without this change, and if the stress test fails I'll try to put the `checkForReadyTasks()` call on a different thread.

**Comment by danberindei** [comment]
@pruivo `WithinThreadExecutor` seems to do the trick for `ReplTotalOrderStateTransferFunctional2PcTest`, I've pushed it.

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java [hunk 2]
```diff
@@ -88,14 +93,15 @@ public void stop() {
...
       runningCaches.put(cacheName, cacheStatus);
 
       long timeout = joinInfo.getTimeout();
       long endTime = timeService.expectedEndTime(timeout, TimeUnit.MILLISECONDS);
-      // Synchronize here to delay any rebalance until after we have received the initial cache topology.
-      // This ensures that the cache will have a topology at the end of startup (with awaitInitialTransfer disabled).
-      synchronized (cacheStatus) {
+      // Pretend the join elay any rebalance until after we have received the initial cache topology.
+      // This ensures that the initial topology and the GET_CACHE_LISTENERS request will happen on the user thread
+      cacheStatus.getTopologyUpdatesCompletionService().continueTaskInBackground();
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: why is this here? the join is not processed inside the SemaphoreCompletionService

**Comment by pruivo** [comment]
@danberindei Maybe it is a little late, but why not using a BlockingRunnable that queues the CacheTopologyCommand until the view is received?

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java (hunk 2 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: I didn't finish updating the comment... I want to make sure the first topology update is on this thread (presumably a user thread), because installing the first topology implies retrieving the clustered listeners and blocking.

**Comment by danberindei** [comment]
@pruivo `GlobalInboundInvocationHandler` always uses `ExecutorService.execute(Runnable)`, which isn't overwritten by `BlockingTaskAwareExecutorServiceImpl`. I didn't want to mess with for now, because it's only half of the problem - I also want to block other updates until the response from join was installed as the initial topology.

**Comment by pruivo** [comment]
two failures in my pc:
```
  DelayedAvailabilityUpdateTest.testDelayedAvailabilityUpdate4:37->testDelayedAvailabilityUpdate:80 Key MagicKey#k1Existing{3a3bed24@DelayedAvailabilityUpdateTest-NodeT-11918/18} available in cache DelayedAvailabilityUpdateTest-NodeR-62688
  EntryRetrieverReplPartitionHandlingTest.testRetrievalWhenPartitionIsDegrated » Test
```

**Comment by danberindei** [comment]
@pruivo I got them on my machine as well, investigating.

**Comment by danberindei** [comment]
@pruivo updated, the tests work for me now. I'm still getting a failure in `CacheManagerTest.testCacheManagerRestartReusingConfigurations` when I run with `taskset -c 1` but I think we never fixed that one properly...

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java [hunk 3]
```diff
@@ -88,14 +94,19 @@ public void stop() {
    public CacheTopology join(String cacheName, CacheJoinInfo joinInfo, CacheTopologyHandler stm,
          PartitionHandlingManager phm) throws Exception {
       log.debugf("Node %s joining cache %s", transport.getAddress(), cacheName);
-      LocalCacheStatus cacheStatus = new LocalCacheStatus(joinInfo, stm, phm);
+
+      // For Total Order caches, we must not move the topology updates to another thread
+      ExecutorService topologyUpdatesExecutor = joinInfo.isTotalOrder() ? new WithinThreadExecutor() : stateTransferExecutor;
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: maybe I'm a little nitpick here but you can have a single WithinThreadExecutor shared

File: core/src/test/java/org/infinispan/stress/LargeCluster2StressTest.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+import java.util.concurrent.Future;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.testng.AssertJUnit.assertEquals;
+
+/**
+ * Test that we're able to start a cluster with lots of caches in a single JVM.
+ *
+ * @author Dan Berindei
+ * @since 5.3
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/stress/LargeCluster2StressTest.java]: 5.3?

**Comment by danberindei** [comment]
@pruivo Updated again.

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java [hunk 5]
```diff
@@ -47,21 +50,27 @@
...
 
+
    @Inject
    public void inject(Transport transport,
                       @ComponentName(ASYNC_TRANSPORT_EXECUTOR) ExecutorService asyncTransportExecutor,
+                      @ComponentName(STATE_TRANSFER_EXECUTOR) ExecutorService stateTransferExecutor,
                       GlobalComponentRegistry gcr, TimeService timeService) {
       this.transport = transport;
       this.asyncTransportExecutor = asyncTransportExecutor;
+      this.stateTransferExecutor = stateTransferExecutor;
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: I got a warning here. the stateTransferExecutor is never used

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java [hunk 6]
```diff
@@ -82,20 +91,26 @@ public void stop() {
          log.tracef("Stopping LocalTopologyManager on %s", transport.getAddress());
       }
       running = false;
+      withinThreadExecutor.shutdown();
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: well, this is not needed, but on other hand, it is no-op anyway

**Comment by pruivo** [comment]
@danberindei you have comments

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java (hunk 6 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: I thought calling shutdown would prevent other tasks from running, but I guess not... I'll leave it there in case we do change the implementation at some point.

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java (hunk 5 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: Removed

**Comment by danberindei** [comment]
Have another look @pruivo!

**Comment by pruivo** [comment]
looks good @danberindei 
going to test it and push it

**Comment by pruivo** [comment]
integrated. thanks @danberindei
